sound/soc/sof/topology.c

```diff
@@ -1303,6 +1351,27 @@ static int sof_widget_ready(struct snd_soc_component *scomp, int index,
 		return ret;
 	}
 
+	/* if there is only one source/sink pin, there is no need to parse source/sink
+	 * pin binding from topology.
+	 */
+	if (swidget->num_source_pins > 1) {
+		ret = sof_parse_pin_binding(swidget, priv, true);
+		if (ret < 0) {
+			dev_err(scomp->dev, "failed to parse component pin tokens for %s\n",
+				w->name);
+			return ret;
+		}
+	}
+
+	if (swidget->num_sink_pins > 1) {
+		ret = sof_parse_pin_binding(swidget, priv, false);
+		if (ret < 0) {
```

Comment:
you'd need to free the swidget->src_pin_binding here

Reply:
on error, swidget->src_pin_binding is not assigned, and the allocated memory is freed within sof_parse_pin_binding